src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,178 @@
+            Node previousNod = currentNode.prev;
+            currentNode.prev = new Node(currentNode.prev, value, currentNode);
+            previousNod.next = currentNode.prev;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isIndexExist(index);
+        Node node = findNodeByIndex(index);
+        return (T) node.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        isIndexExist(index);
+        Node node = findNodeByIndex(index);
+        T oldValue = (T) node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        isIndexExist(index);
+        T value = (T) findNodeByIndex(index).value;
+        unlink(findNodeByIndex(index));
+        return value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node node = head;
+        for (int i = 0; i < size; i++) {
+            if (needToRemove(object, node)) {
+                unlink(node);
+                return true;
+            }
+            node = node.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> next;
+        private Node<T> prev;
+
+        private Node(Node<T> prev, T element, Node<T> next) {
+            this.value = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    private void isIndexExist(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    private boolean needToRemove(T object, Node node) {
+        return (object == null && node.value == null)
+                || (node.value != null && node.value.equals(object));
+    }
+
+    private Node findNodeByIndex(int index) {
+        if (index == 0) {
+            return head;
+        }
+        if (index == size) {
+            return tail;
+        }
+        if (size / 2 > index) {
+            return findFromHead(index);
+        } else {
+            return findFromTail(index);
+        }
+    }
+
+    private Node findFromHead(int index) {
+        Node currentNode = head;
+        for (int i = 0; i != index; i++) {
+            currentNode = currentNode.next;
+        }
+        return currentNode;
+    }
+
+    private Node findFromTail(int index) {
+        Node currentNode = tail;
+        for (int i = size - 1; i != index; i--) {
+            currentNode = currentNode.prev;
+        }
+        return currentNode;
+    }
+
+    private void unlink(Node node) {
```

Comment:
* size-- can come out of those `if`s to the end of the method to not duplicate it
* there is no point in using both `else` and `return`